### PR TITLE
Laboratory work #3, Egor Kostin - 22FPL2

### PR DESCRIPTION
**Pull request by egorkostin24**

### CODE REVIEW

File: lab_3_generate_by_ngrams/main.py [hunk 1]
```diff
@@ -23,7 +23,8 @@ def __init__(self, end_of_word_token: str) -> None:
         Args:
             end_of_word_token (str): A token denoting word boundary
         """
-
+        self._end_of_word_token = end_of_word_token
+        self._storage = {end_of_word_token: 0}
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:27]: Add a new empty line

File: lab_3_generate_by_ngrams/main.py [hunk 2]
```diff
@@ -34,13 +35,30 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
 
         Args:
             text (str): Original text
-
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:37]: revert

File: lab_3_generate_by_ngrams/main.py [hunk 3]
```diff
@@ -34,13 +35,30 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
+        self.text = text
+        splited_text = text.split()
+        list_of_tokens = []
+        for word in splited_text:
+            current_word = ''.join(char.lower() for char in word if char.isalpha())
+            if current_word:
+                list_of_tokens += current_word
+                list_of_tokens.append(self._end_of_word_token)
+
+        if text[-1].isalnum():
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:55]: why do you do it?

File: lab_3_generate_by_ngrams/main.py [hunk 4]
```diff
@@ -55,15 +73,18 @@ def get_id(self, element: str) -> Optional[int]:
         In case of corrupt input arguments or arguments not included in storage,
         None is returned
         """
+        if not isinstance(element, str) or element not in self._storage:
+            return None
 
+        return self._storage[element]
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:79]: Add a new empty line

File: lab_3_generate_by_ngrams/main.py [hunk 5]
```diff
@@ -55,15 +73,18 @@ def get_id(self, element: str) -> Optional[int]:
...
     def get_end_of_word_token(self) -> str:
         """
         Retrieve value stored in self._end_of_word_token attribute.
 
         Returns:
             str: EoW token
         """
-
+        return self._end_of_word_token
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:87]: Add a new empty line

File: lab_3_generate_by_ngrams/main.py [hunk 6]
```diff
@@ -76,7 +97,12 @@ def get_token(self, element_id: int) -> Optional[str]:
 
         In case of corrupt input arguments or arguments not included in storage, None is returned
         """
+        if not isinstance(element_id, int) or element_id not in self._storage.values():
+            return None
+
+        inverted_storage = {v: k for k, v in self._storage.items()}
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:103]: You can do less iteration if return from the loop a needed element

File: lab_3_generate_by_ngrams/main.py [hunk 7]
```diff
@@ -104,7 +149,11 @@ def _put(self, element: str) -> None:
         In case of corrupt input arguments or invalid argument length,
         an element is not added to storage
         """
+        if not isinstance(element, str) or len(element) != 1:
+            return None
 
+        if element not in self._storage:
+            self._storage[element] = len(self._storage)
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:156]: Add a new empty line

File: lab_3_generate_by_ngrams/main.py [hunk 8]
```diff
@@ -93,7 +119,26 @@ def encode(self, text: str) -> Optional[tuple[int, ...]]:
...
+        for token in tokenized_text:
+            self._put(token)
+            encoded_numbers = self.get_id(token)
 
+            if encoded_numbers is None:
+                return None
+
+            encoded_corpus.append(encoded_numbers)
+
+        return tuple(encoded_corpus)
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:141]: Add a new empty line

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py (hunk 3 above)
Comment (egorkostin24) [reply lab_3_generate_by_ngrams/main.py:55]: I have not fully understood the question, unfortunately. I first made a list from the text with the split method, then made all words lower if they consist of letters and added them to the empty list, appending the end of word token...
If that's what you were asking about....

File: lab_3_generate_by_ngrams/main.py (hunk 1 above)
Comment (egorkostin24) [reply lab_3_generate_by_ngrams/main.py:28]: done

File: lab_3_generate_by_ngrams/main.py (hunk 2 above)
Comment (egorkostin24) [reply lab_3_generate_by_ngrams/main.py]: done

File: lab_3_generate_by_ngrams/main.py (hunk 4 above)
Comment (egorkostin24) [reply lab_3_generate_by_ngrams/main.py:82]: done

File: lab_3_generate_by_ngrams/main.py (hunk 5 above)
Comment (egorkostin24) [reply lab_3_generate_by_ngrams/main.py:91]: done

File: lab_3_generate_by_ngrams/main.py (hunk 7 above)
Comment (egorkostin24) [reply lab_3_generate_by_ngrams/main.py:163]: done

**Pull request closed without merging**